krun/platform.py

```diff
@@ -815,6 +845,93 @@ def get_reboot_cmd(self):
         cmd.append(self.REBOOT)
         return cmd
 
+    def pin_process_args(self):
+        """Pin to a set of adaptive tick CPUs.
+        We are working the assumption that the kernel is NO_HZ_FULL_ALL meaning
+        that all but the first CPU are in adaptive tick mode."""
+
+        if self.num_cpus == 1:
+            fatal("not enough CPUs to pin")
+
+        cpus = ",".join([str(x) for x in xrange(1, self.num_cpus)])
+        return ["taskset", "-c", cpus]
+
+    def _check_util_linux_installed(self):
+        from distutils.spawn import find_executable
+        if not find_executable("taskset"):
+            fatal("util-linix is not installed "
+                  "(needed for pinning and scheduler tweaking).")
+
+    def sanity_checks(self):
+        UnixLikePlatform.sanity_checks(self)
+        self._check_util_linux_installed()
+        self._check_isolcpus()
+        self._sanity_check_cpu_affinity()
+        self._sanity_check_scheduler()
+
+    def _sanity_check_cpu_affinity(self):
+        from krun.vm_defs import NativeCodeVMDef
+        from krun import EntryPoint
+
+        ep = EntryPoint("check_linux_cpu_affinity.so")
+        vd = NativeCodeVMDef()
+        util.spawn_sanity_check(self, ep, vd, "CPU affinity",
+                                force_dir=PLATFORM_SANITY_CHECK_DIR)
+
+    def _sanity_check_scheduler(self):
+        from krun.vm_defs import NativeCodeVMDef
+        from krun import EntryPoint
+
+        ep = EntryPoint("check_linux_scheduler.so")
+        vd = NativeCodeVMDef()
+        util.spawn_sanity_check(self, ep, vd, "Scheduler",
+                                force_dir=PLATFORM_SANITY_CHECK_DIR)
+
+    def _advise_isolcpus_arg(self, got_cpus, expect_cpus):
+        """Error out and guide user in isolating CPUs"""
+
+        arg = "isolcpus=%s" % ",".join(expect_cpus)
+        self._fatal_kernel_arg(
+            arg, "CPUs incorrectly isolated. Got: %s, expect: %s" %
+            (got_cpus, expect_cpus)
+        )
+
+    def _check_isolcpus(self):
+        """Checks the correct CPUs have been isolated.
+        (All but the boot processor)"""
+
+        expect_cpus = [str(x) for x in xrange(1, self.num_cpus)]
+        all_args = self._get_kernel_cmdline()
+
+        args = all_args.split(" ")
+        for arg in args:
+            if '=' not in arg:
+                continue
```

Comment:
I never know what to use tbh. It's mixed in every python project i ever worked on.